src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -1413,7 +1417,114 @@ private IBlockOperation CreateBoundBlockOperation(BoundBlock boundBlock)
+                            constantValue,
+                            isImplicit: true);
+                    }
+
+                    var argumentNullExceptionMethodSymbol = (IMethodSymbol)compilation.GetWellKnownTypeMember(WellKnownMember.System_ArgumentNullException__ctorString);
+                    var argumentNullExceptionType = compilation.GetWellKnownType(WellKnownType.System_ArgumentNullException);
+                    var argumentNullExceptionObject = new ObjectCreationOperation(
+                                constructor: argumentNullExceptionMethodSymbol,
+                                initializer: new ObjectOrCollectionInitializerOperation(
+                                        initializers: ImmutableArray<IOperation>.Empty,
+                                        _semanticModel,
+                                        syntax,
+                                        compilation.GetWellKnownType(WellKnownType.System_ArgumentNullException),
+                                        constantValue,
+                                        isImplicit: true),
+                                ImmutableArray.Create<IArgumentOperation>(
+                                    new ArgumentOperation(
+                                        new LiteralOperation(_semanticModel, syntax, compilation.GetSpecialType(SpecialType.System_String), param.Name, isImplicit: true),
+                                        ArgumentKind.Explicit,
+                                        parameter: argumentNullExceptionMethodSymbol.Parameters[0],
+                                        inConversionOpt: null,
+                                        outConversionOpt: null,
+                                        _semanticModel,
+                                        syntax,
+                                        isImplicit: true)),
+                                _semanticModel,
+                                syntax,
+                                type: argumentNullExceptionType,
+                                constantValue,
+                                isImplicit: true);
+
+                    IOperation whenTrue = new ExpressionStatementOperation(
+                        new ThrowOperation(
+                            argumentNullExceptionObject,
+                            _semanticModel,
+                            syntax,
+                            type: argumentNullExceptionType,
+                            constantValue,
+                            isImplicit: true),
+                        _semanticModel,
+                        syntax,
+                        compilation.GetSpecialType(SpecialType.System_Void),
+                        constantValue,
+                        isImplicit: true);
+                    var newIf = new ConditionalOperation(conditionOp, whenTrue, null, isRef: false, _semanticModel, syntax, compilation.GetSpecialType(SpecialType.System_Boolean), constantValue, isImplicit: true);
```

Comment:
>var newIf = new ConditionalOperation [](start = 20, length = 36)

Consider extracting the calculation of this parameter check to a separate method so that the loop over the parameters is separated from the transformation of each parameter. (It seems like the two are separable.)